[Performance](Variant) Improve load performance for variant type

eldenmoon opened the pull request:
1. remove phmap for padding rows
2. add `SimpleFieldVisitorToScarlarType` for short circuit type deducing
3. correct type coercion for conflict types bettween integers
4. improve nullable column performance
5. remove shared_ptr dependancy for DataType use TypeIndex instead
6.  Optimization by caching the order of fields (which is almost always the same)
    and a quick check to match the next expected field, instead of searching the hash table.

In clickbench data, load performance:
12m36.799s ->7m10.934s about 43% latency reduce

In variant_p2/performance.groovy:
3min44s20 -> 1min15s80 about 66% latency reducy
 
## Proposed changes

Issue Number: close #xxx

<!--Describe your changes.-->

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

eldenmoon committed:
[Performance](Variant) Improve load performance for variant type

1. remove phmap for padding rows
2. add `SimpleFieldVisitorToScarlarType` for short circuit type deducint
3. correct type coercion for conflict types bettween integers

In clickbench data, load performance:
12m36.799s -> 8m30.870s about 32.5%

In variant_p2/performance.groovy:
3min44s20 -> 2min21s07 about 36.7%

eldenmoon commented:
run buildall

airborne12 reviewed (approved):
LGTM

eldenmoon committed:
fix

eldenmoon commented:
run buildall

eldenmoon committed:
fix and remove shared_ptr dependancy for DataType use TypeIndex instead

eldenmoon commented:
run buildall

eldenmoon commented:
run buildall

eldenmoon commented:
run p0

xiaokang commented on be/src/vec/common/schema_util.cpp:
```diff
@@ -114,31 +114,8 @@ bool is_conversion_required_between_integers(const IDataType& lhs, const IDataTy
-        return !(rhs == FieldType::OLAP_FIELD_TYPE_TINYINT);
-    }
-    return true;
+    return (!is_native_int && !is_native_uint) ||
```
opposite to original code?

eldenmoon replied:
yes， the original code logical is wrong

xiaokang commented on be/src/vec/columns/column_object.cpp:
```diff
@@ -148,6 +151,84 @@ class FieldVisitorToNumberOfDimensions : public StaticVisitor<size_t> {
+// Visitor that allows to get type of scalar field
+// but include contain complex field.This is a faster version
+// for FieldVisitorToScalarType which does not support complex field.
+class SimpleFieldVisitorToScarlarType : public StaticVisitor<size_t> {
```
typo SimpleFieldVisitorToScalarType

eldenmoon replied:
fixed

xiaokang commented on be/src/vec/columns/column_object.cpp:
```diff
@@ -308,18 +399,29 @@ void ColumnObject::Subcolumn::insert(Field field, FieldInfo info) {
+                                                                 *nested_common_base_type)) {
+            LOG_EVERY_N(INFO, 100) << "Conversion between " << base_type->get_name() << " and "
+                                   << least_common_base_type->get_name();
             get_least_supertype<LeastSupertypeOnError::Jsonb>(
```
Is least_common_base_type changed after calling get_least_supertype? If no, where is get_least_supertype updated?

eldenmoon replied:
in `add_new_column_part`

eldenmoon commented:
run buildall

eldenmoon committed:
Optimization by caching the order of fields (which is almost always the same)
and a quick check to match the next expected field, instead of searching the hash table.

eldenmoon commented:
run buildall

eldenmoon committed:
Merge branch 'master' into var-perf

eldenmoon commented:
run buildall

xiaokang commented on be/src/vec/columns/column_object.cpp:
```diff
@@ -220,20 +281,30 @@ class FieldVisitorToScalarType : public StaticVisitor<size_t> {
+    if (field.is_complex_field()) {
+        get_field_info_impl<FieldVisitorToScalarType>(field, info);
+    } else {
+        get_field_info_impl<SimpleFieldVisitorToScalarType>(field, info);
```
Is there any opt for SimpleFieldVisitorToScalarType?
